### PR TITLE
feat(form-plugin): implement `NgxsFormPluginValueChangesStrategy`

### PR DESCRIPTION
**Pull request by arturovt**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/ngxs/store/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

```
[x] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] Other... Please describe:
```

## What is the current behavior?

Issue Number: #429 

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

## Other information

Currently our users don't have an access to the `valueChanges` stream that is listened under the hood and followed by dispatching 3 actions `UpdateFormValue`, `UpdateFormDirty`, `UpdateFormErrors`. `ngxsFormDebounce` binding provides the ability to create a timeout between dispatched actions.

This PR is very lightweight (not a lot code was added) and introduces a mechanism that is called "value changes strategy". The "noop" strategy, that doesn't change the behavior, is used by default. You can look at the `DeepEquals` strategy that pipes `valueChanges` stream with `distinctUntilChanged` operator and passes a callback that serializes previous and current value, and will generate event only if they differ. Thus we can reduce the amount of dispatched actions and give users the ability to do anything with `valueChanges` stream before actions are dispatched and state changed.

API:
```html
<form ngxsForm="..." [ngxsFormValueChangesStrategy]="ClassReference"> ... </form>
```

Notice that `ngxsFormValueChangesStrategy` binding awaits the class reference and not the class instance. Thus its instance can be retrieved lazily.

The same approach is used inside the `ngComponentOutlet` directive:
```ts
@Component({selector: 'hello-world', template: 'Hello World!'})
export class HelloWorld {
}

@Component({
  selector: 'ng-component-outlet-simple-example',
  template: `<ng-container *ngComponentOutlet="HelloWorld"></ng-container>`
})
export class NgTemplateOutletSimpleExample {
  // This field is necessary to expose HelloWorld to the template.
  HelloWorld = HelloWorld;
}
```

---

Strategies are provided on the directive level and not globally.  E.g. I'm using the `ngxsForm` directive at  10 places. I want to provide the `DeepEqualNgxsFormPluginValueChangesStrategy` only at 2 places.

---

Basically any developer can do ANY stuff with the underneath form as it gives an access to the `valueChanges` stream.

### CODE REVIEW

**Comment by arturovt** [comment]
@splincode umm, hard to test, I thought if any test is broken then it did OK. :D

**Comment by splincode** [comment]
@arturovt current behavior `state changes every time a form action is dispatched`, you can send PR with calculate count state changes before current pull request and update this PR with new behavior on tests

**Comment by arturovt** [comment]
@splincode 

> calculate count state changes 

Dispatching 1 action instead of 3 is 3 three times less :smile:

**Comment by splincode** [comment]
@arturovt This does not violate backward compatibility? 

```ts
import { UpdateFormDirty } from '@ngxs/form-plugin';

class MyComponent {
 constructor(private actions$: Actions) {}
 public ngOnInit(actions$: Action) {
   this.actions$.pipe(ofActionSuccessful(UpdateFormDirty)).subscribe(() => alert('update dirty'));
 }
}
```

**Comment by arturovt** [comment]
@splincode yeah, you're right. Still need to discuss how we can reduce it

**Comment by arturovt** [comment]
@splincode what do you think about `distinctUntilChanged` that compares objects?

File: packages/form-plugin/src/directive.ts [hunk 1]
```diff
@@ -32,6 +39,11 @@ export class FormDirective implements OnInit, OnDestroy {
   ) {}
 
   ngOnInit() {
+    if (this.useDefaultCompare) {
+      // Create it lazily by condition
+      this.compare = (a, b) => JSON.stringify(a) === JSON.stringify(b);
```
Comment (splincode) [inline packages/form-plugin/src/directive.ts]: I think is very hard

```ts
JSON.stringify(a) === JSON.stringify(b)
```

File: packages/form-plugin/tests/value-changes-strategy.spec.ts [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+
+    subscription.unsubscribe();
+
+    // Assert
+    expect(dispatchedTimes).toBe(6);
+  });
+
+  it('should pipe via "distinctUntilChanged" if custom strategy provided with such operator', () => {
+    // Arrange
+    class CustomNgxsFormPluginValueChangesStrategy
```
Comment (splincode) [inline packages/form-plugin/tests/value-changes-strategy.spec.ts]: I think you can make it public api.

```CustomNgxsFormPluginValueChangesStrategy``` -> ```NgxsFormPluginDeepEqualsStrategy```

example for people

```ts
import { NgxsFormPluginDeepEqualsStrategy } from '@ngxs/form-plugin';

NgxsFormPluginModule.forRoot({
    valueChangesStrategy: NgxsFormPluginDeepEqualsStrategy
})
```

**Comment by markwhitfeld** [comment]
@arturovt Could you please describe the problem that this solves and what the new behavior provides in the PR description. Just a link to the original issue does not give enough information on what this is supposed to do.

**Comment by splincode** [comment]
@arturovt Update docs in current PR

**Comment by arturovt** [comment]
> @arturovt Could you please describe the problem that this solves and what the new behavior provides in the PR description. Just a link to the original issue does not give enough information on what this is supposed to do.

Sure. As I told you I will add descriptions for next PRs.

Currently our users don't have an access to the `valueChanges` stream that is listened under the hood and followed by dispatching 3 actions `UpdateFormValue`, `UpdateFormDirty`, `UpdateFormErrors`. `ngxsFormDebounce` binding provides the ability to create a timeout between dispatched actions.

This PR is very lightweight (not a lot code was added) and introduces a strategy that users can apply to the `valueChanges` stream. The "noop" strategy, that doesn't change the behavior, is used by default. You can look at the `DeepEquals` strategy that pipes `valueChanges` stream with `distinctUntilChanged` operator and passes a callback that serializes previous and current value, and will generate event only if they differ. Thus we can reduce the amount of dispatched actions and give users the ability to do anything with `valueChanges` stream before actions are dispatched and state changed.

**Comment by splincode** [comment]
@arturovt @markwhitfeld  any updates?

File: packages/form-plugin/src/directive.ts [hunk 3]
```diff
@@ -25,15 +36,22 @@ export class FormDirective implements OnInit, OnDestroy {
   @Input('ngxsFormClearOnDestroy')
   clearDestroy = false;
 
+  /**
+   * `valueChangesStrategy` is noop by default but can be overwritten via binding
+   */
+  @Input('ngxsFormValueChangesStrategy')
+  valueChangesStrategy: Type<
+    NgxsFormPluginValueChangesStrategy
+  > = NoopNgxsFormPluginValueChangesStrategy;
```
Comment (splincode) [inline packages/form-plugin/src/directive.ts]: And if I get tired every time for each form to redefine the strategy?

File: packages/form-plugin/tests/value-changes-strategy.spec.ts [hunk 4]
```diff
@@ -1,19 +1,19 @@
...
 import {
   NgxsFormPluginModule,
   UpdateFormValue,
   NgxsFormPluginValueChangesStrategy,
   DeepEqualNgxsFormPluginValueChangesStrategy
 } from '..';
+import { FormDirective } from '../src/directive';
 
-describe('NgxsFormPluginValueChangesStrategy', () => {
+fdescribe('NgxsFormPluginValueChangesStrategy', () => {
```
Comment (splincode) [inline packages/form-plugin/tests/value-changes-strategy.spec.ts]: `fdescribe` instead `describe`?

File: packages/form-plugin/src/directive.ts (hunk 3 above)
Comment (arturovt) [reply packages/form-plugin/src/directive.ts]: Your suggestions?

Comment (splincode) [reply packages/form-plugin/src/directive.ts]: support 

```ts
NgxsFormPlugin.forRoot({ defaultStategy: NoopNgxsFormPluginValueChangesStrategy })
```

Comment (arturovt) [reply packages/form-plugin/src/directive.ts]: What is this for? Noop strategy is used by default.

Comment (splincode) [reply packages/form-plugin/src/directive.ts]: But if I want to make a different strategy in the application for all forms at once

```ts
NgxsFormPlugin.forRoot({ defaultStategy: DeepEqualNgxsFormPluginValueChangesStrategy })
```

or 

```ts
NgxsFormPlugin.forRoot({ defaultStategy: CustomStrategy })
```

Comment (arturovt) [reply packages/form-plugin/src/directive.ts]: That makes sense.

Tho to avoid making changes and then revert it back, let's wait for @markwhitfeld response. If it's OK then I will add this option.

**Comment by arturovt** [comment]
@markwhitfeld 

The reason to add this feature is to let developers access the `valueChanges` property before dispatching those 3 actions.

> that could just be a simple setting on the directive

I'm open to any other solutions. Really. I do not insist on this "value changes strategy" so if you'd offer something easier - that would be great.

---

The first use case is that "deep equal strategy" that's provided in the PR. `distinctUntilChanged` is basically a "standard" today. By "standard' I mean that it's on of the most popular operators used in ~100% applications I guess.

Assume we've got a form with `username` and `email` controls and we want to check their uniqueness:

```ts
@Component({
  template: `
    <form [formGroup]="form">
      <input formControlName="username">
      <input formControlName="email">
      <p *ngIf="form.hasError('unique')">Values are not unique</form>
    </form>
  `
})
export class FormComponent {
  form = new FormGroup({
    username: new FormControl(),
    email: new FormControl()
  });

  constructor(private uniqueService: UniqueService) {
    this.setupListener();
  }

  private setupListener(): void {
    this.form.valueChanges
      .pipe(
        debounceTime(500),
        distinctUntilChanged((a, b) => JSON.stringify(a) === JSON.stringify(b)),
        switchMap(value => this.uniqueService.checkIfUsernameAndEmailAreUnique(value))
      )
      .subscribe(unique => {
        this.form.setErrors({ unique });
      });
  }
}
```

Async validator is not suitable here as we can't debounce it.

---

The second use case would be adding some "stream projection" operator like `switch|concatMap` as in the example above. This could be search for example among different predicates (for example name/start date/end date). I want to synchronize those predicates with my state also to save to local storage for example and to show the user next time:

```ts
@Component({
  template: `
    <form [formGroup]="form">
      <mat-form-field>
        <input matInput formControlName="name" />
      </mat-form-field>

      <mat-form-field>
        <input matInput [matDatepicker]="startDatePicker" formControlName="startDate" />
        <mat-datepicker-toggle matSuffix [for]="startDatePicker"></mat-datepicker-toggle>
        <mat-datepicker #startDatePicker></mat-datepicker>
      </mat-form-field>

      <mat-form-field>
        <input matInput [matDatepicker]="endDatePicker" formControlName="endDate" />
        <mat-datepicker-toggle matSuffix [for]="endDatePicker"></mat-datepicker-toggle>
        <mat-datepicker #endDatePicker></mat-datepicker>
      </mat-form-field>
    </form>
  `
})
export class FormComponent {
  form = new FormGroup({
    name: new FormControl(),
    startDate: new FormControl(),
    endDate: new FormControl()
  });

  constructor(private searchService: SearchService) {
    this.setupListener();
  }

  private setupListener(): void {
    this.form.valueChanges
      .pipe(
        distinctUntilChanged((a, b) => JSON.stringify(a) === JSON.stringify(b)),
        switchMap(value => this.searchService.search(value))
      )
      .subscribe(items => {
        this.store.dispatch(new UpdateFormValue(...))
      });
  }
}
```

---

So my point is that developers have to be able to access the `valueChanges` stream under the hood before dispatching 3 actions. As even my logger plugin tears apart when I put some letter into some input as it prints 3 actions immediately.

**Comment by splincode** [comment]
@arturovt I think I know how it would be better, I will write an example

**Comment by markwhitfeld** [comment]
I see what you are getting at, but the developer has access to the `valueChanges` stream on the form anyway. I think that the `distinctUntilChanged` approach using deep equality should be a simple attribute for the directive.
Something like:
```
<form [formGroup]="newNovelForm" ngxsForm="novels.newNovelForm" ngxsFormDistinct="deepEquality" >
```
This would expect the string "deepEquality".
What do you think?

**Comment by arturovt** [comment]
@markwhitfeld 

> I see what you are getting at, but the developer has access to the `valueChanges` stream on the form anyway.

He has. But this stream is subscribed separately under the hood.

> I think that the distinctUntilChanged approach using deep equality should be a simple attribute for the directive

`distinctUntilChanged` was used for demonstration purposes. You can have the ability to apply ANY set of operators.

**Comment by splincode** [comment]
@markwhitfeld @arturovt any progress?

**Comment by arturovt** [comment]
> @markwhitfeld @arturovt any progress?

> @arturovt I think I know how it would be better, I will write an example

**Comment by splincode** [comment]
@arturovt not idea

**Comment by splincode** [comment]
Maybe, you can

```html
<form [formGroup]="newNovelForm" ngxsForm="novels.newNovelForm" ngxsFormDistinct="deepEquality" .. />
```

**Comment by splincode** [comment]
Why closed?((
